security_privacy_ln.asciidoc

```diff
@@ -0,0 +1,614 @@
+
+=== DoS in Lightning
+
+Similarly to Bitcoin, the Lightning Network charges fees for the use of its public resources, but in this case, the resources are public channels, and the fees come in the form of routing fees.
+The ability to route payments through nodes in exchange for fees provides the network with a large scalability benefit - nodes that are not directly connected can still transact - but it comes at the cost of exposing a public resource that must be protected against DoS attacks.
+
+When a Lightning node forwards a payment on your behalf, it uses bandwidth to update its commitment transaction, and the amount of the payment is reserved in their channel balance until it is settled or failed.
+In successful payments, this is acceptable because the node is eventually paid out its fees.
+Failed payments do not incur fees in the current protocol.
+This allows nodes to costlessly route failed payments through any channels.
+This is great for legitimate users, who wouldn't like to pay for failed attempts, but also allows attackers to costlessly consume nodes' resources - much like the low-fee transactions on Bitcoin that never end up paying miner fees.
+
+At the time of writing, a discussion is https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-June/002734.html[ongoing] on the lightning-dev mailing list as to how best address this issue.
+
+=== Known DoS attacks
+
+There are two known DoS attacks on public Lightning channels which render a target channel, or a set of target channels, unusable.
+Both attacks involve routing payments through a public channel, then holding them until their timeout, thus maximizing the attack's duration.
+The requirement to fail payments to not pay fees is fairly simple to meet because malicious nodes can simply reroute payments to themselves.
+In the absence of fees for failed payments, the only cost to the attacker is the on-chain cost of opening a channel to dispatch these payments through, which can be trivial in low fee environments.
+
+==== Commitment jamming
+
+Lightning nodes update their shared state using asymmetric commitment transactions, on which HTLCs are added and removed to facilitate payments.
+Each party is limited to a total of https://github.com/lightningnetwork/lightning-rfc/blob/c053ce7afb4cbf88615877a0d5fc7b8dbe2b9ba0/02-peer-protocol.md#the-open_channel-message[483] on the commitment at a time.
+A channel jamming attack allows an attacker to render a channel unusable by routing 483 payments through the target channel and holding them until they timeout.
+
+It should be noted that this limit is arbitrary and was chosen in the specification to ensure that all the HTLCs can be swept in a https://github.com/lightningnetwork/lightning-rfc/blob/master/05-onchain.md#penalty-transaction-weight-calculation[single justice transaction].
+While this limit _may_ be increased, transactions are still limited by the block size, so the number of slots available is likely to remain limited.
```

Comment:
One missing component here is that it's possible to essentially lift this limit (by using indirect commitments), instead letting nodes manage the free HTLC slots as if they were space in a queue. Modern implementations also feature something commonly referred to as an "HTLC interceptor" that allows them to make accept/deny decisions based on the current channel state. In a sense, the problem is similar to TCP congestion control/avoidance and buffer bloat itself.